File: cli/src/main/java/de/mobanisto/lanchat/ReceiveFile.java
package de.mobanisto.lanchat;

import static java.nio.file.StandardOpenOption.CREATE;
import static java.nio.file.StandardOpenOption.TRUNCATE_EXISTING;
import static java.nio.file.StandardOpenOption.WRITE;

import java.io.IOException;
import java.io.InputStream;
import java.net.ServerSocket;
import java.net.Socket;
import java.nio.channels.Channels;
import java.nio.channels.FileChannel;
import java.nio.channels.ReadableByteChannel;
import java.nio.file.Path;
import java.nio.file.Paths;

public class ReceiveFile
{

	public static void main(String[] args) throws IOException
	{
		if (args.length != 1) {
			System.out.println("usage: receive-file <file>");
			System.exit(1);
		}

		String argOutput = args[0];
		Path output = Paths.get(argOutput);

		try (ServerSocket serverSocket = new ServerSocket(5001)) {
			System.out.println("Waiting for incoming file...");
			Socket socket = serverSocket.accept();
			System.out.println("Receiving file...");
			InputStream input = socket.getInputStream();
			try (FileChannel channel = FileChannel.open(output, CREATE,
					TRUNCATE_EXISTING, WRITE);
					ReadableByteChannel inputChannel = Channels
							.newChannel(input);) {
				long transferred = channel.transferFrom(inputChannel, 0,
						Integer.MAX_VALUE);
				System.out.println(
						String.format("Received %d bytes", transferred));
			}
		}
	}

}

File: desktop/src/main/kotlin/de/mobanisto/lanchat/ReceiveFile.kt
package de.mobanisto.lanchat

import com.github.ajalt.clikt.core.CliktCommand
import com.github.ajalt.clikt.parameters.arguments.argument
import com.github.ajalt.clikt.parameters.options.default
import com.github.ajalt.clikt.parameters.options.option
import com.github.ajalt.clikt.parameters.types.int
import com.github.ajalt.clikt.parameters.types.path
import java.net.ServerSocket
import java.nio.channels.Channels
import java.nio.channels.FileChannel
import java.nio.file.StandardOpenOption
import kotlin.io.path.absolute

class ReceiveFile : CliktCommand(name = "receive-file") {

    private val port: Int by option(
        "--port",
        help = "Port to receive files on"
    ).int().default(5001)

    private val output by argument().path()

    override fun run() {
        println(output.absolute())
        ServerSocket(port).use { serverSocket ->
            println("Waiting for incoming file...")
            val socket = serverSocket.accept()
            println("Receiving file...")
            val input = socket.getInputStream()
            FileChannel.open(
                output,
                StandardOpenOption.CREATE,
                StandardOpenOption.TRUNCATE_EXISTING,
                StandardOpenOption.WRITE
            ).use { channel ->
                Channels.newChannel(input).use { inputChannel ->
                    val transferred = channel.transferFrom(inputChannel, 0, Int.MAX_VALUE.toLong())
                    println(String.format("Received %d bytes", transferred))
                }
            }
        }
    }
}
